email/reporter.go

```diff
@@ -0,0 +1,231 @@
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+const (
+	// reportingFrequency defines the frequency with which we scan the database
+	// to report csam reports to NCMEC
+	reportingFrequency = 30 * time.Second
+
+	// maxShutdownTimeout is the amount of time we wait on the waitgroup when
+	// Stop is being called before returning an error that indicates an unclean
+	// shutdown.
+	maxShutdownTimeout = time.Minute
+)
+
+type (
+	// Reporter is an object that will periodically scan the database for CSAM
+	// abuse reports that have not been reported to NCMEC yet.
+	Reporter struct {
+		staticClient    *NCMECClient
+		staticDatabase  *database.AbuseScannerDB
+		staticLogger    *logrus.Entry
+		staticStopChan  chan struct{}
+		staticWaitGroup sync.WaitGroup
+	}
+)
+
+// NewReporter creates a new reporter.
+func NewReporter(database *database.AbuseScannerDB, creds NCMECCredentials, logger *logrus.Logger) *Reporter {
+	return &Reporter{
+		staticClient:   NewNCMECClient(creds),
+		staticDatabase: database,
+		staticLogger:   logger.WithField("module", "Reporter"),
+		staticStopChan: make(chan struct{}),
+	}
+}
+
+// Start initializes the reporter process.
+func (r *Reporter) Start() error {
+	r.staticWaitGroup.Add(1)
+	go func() {
+		r.threadedReportMessages()
+		r.staticWaitGroup.Done()
+	}()
+	return nil
+}
+
+// Stop waits for the finalizer's waitgroup and times out after one minute.
+func (r *Reporter) Stop() error {
+	close(r.staticStopChan)
+
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		r.staticWaitGroup.Wait()
+	}()
+	select {
+	case <-c:
+		return nil
+	case <-time.After(maxShutdownTimeout):
+		return errors.New("unclean reporter shutdown")
+	}
+}
+
+// threadedReportMessages will periodically fetch email messages that have not
+// been tagged as csam and have not been reported to NCMEC yet.
+func (r *Reporter) threadedReportMessages() {
+	// convenience variables
+	logger := r.staticLogger
+
+	// create a new ticker
+	ticker := time.NewTicker(reportingFrequency)
+
+	// start the loop
+	for {
+		logger.Debugln("threadedReportMessages loop iteration triggered")
+		r.reportMessages()
+
+		select {
+		case <-r.staticStopChan:
+			logger.Debugln("Reporter stop channel closed")
+			return
+		case <-ticker.C:
+		}
+	}
+}
+
+// reportMessages fetches all unreported messages from the database and reports
+// them to NCMEC. Reporting only applies to emails which have been tagged with
+// CSAM, we report those messages and the content we find inside them to NCMEC.
+func (r *Reporter) reportMessages() {
+	// convenience variables
+	abuseDB := r.staticDatabase
+	logger := r.staticLogger
+
+	// fetch all unreported emails
+	toReport, err := abuseDB.FindUnreported()
+	if err != nil {
+		logger.Errorf("Failed fetching unreported emails, error %v", err)
+		return
+	}
+
+	// log unreported message count
+	numUnreported := len(toReport)
+	if numUnreported == 0 {
+		logger.Debugf("Found %v unreported messages", numUnreported)
```

Comment:
Could do this before the `if` to avoid calling it in 2 places.

Reply:
There are two different log levels, if it's fine by you I'd like to keep it the way it is.

If the abuse scanner is run in `debug` mode the modules log a message indicating they didn't found any "work".
If the abuse scanner is run in `info` mode it will only log a message when it's about to process work.